nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterAttribute.java

```diff
@@ -0,0 +1,273 @@
+        _propertyDescriptors.add(MATCHING_STRATEGY);
+        _propertyDescriptors.add(ATTRIBUTE_SET);
+        _propertyDescriptors.add(ATTRIBUTE_REGEX);
+        properties = Collections.unmodifiableList(_propertyDescriptors);
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    private volatile Predicate<String> cachedMatchingPredicate;
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+        final MatchingStrategy matchingStrategy = getMatchingStrategy(context);
+
+        cachedMatchingPredicate = null;
+
+        if (matchingStrategy == MatchingStrategy.ENUMERATION
+                && !context.getProperty(ATTRIBUTE_SET).isExpressionLanguagePresent()) {
+            cachedMatchingPredicate = determineMatchingPredicateBasedOnEnumeration(context, null);
+        }
+        if (matchingStrategy == MatchingStrategy.REGEX
+                && !context.getProperty(ATTRIBUTE_REGEX).isExpressionLanguagePresent()) {
+            cachedMatchingPredicate = determineMatchingPredicateBasedOnRegex(context, null);
+        }
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        final FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        final Predicate<String> matchingPredicate = determineMatchingPredicate(context, flowFile);
+
+        final FilterMode filterMode = getFilterMode(context);
+        final Predicate<String> isMatched;
+        if (filterMode == FilterMode.RETAIN) {
+            isMatched = matchingPredicate;
+        } else {
+            isMatched = matchingPredicate.negate();
+        }
+
+        final Set<String> attributesToRemove = new HashSet<>(flowFile.getAttributes().keySet());
+        attributesToRemove.removeIf(isMatched);
+
+        final FlowFile updatedFlowFile = session.removeAllAttributes(flowFile, attributesToRemove);
+        session.transfer(updatedFlowFile, REL_SUCCESS);
+    }
+
+    private Predicate<String> determineMatchingPredicate(ProcessContext context, FlowFile flowFile) {
+        if (cachedMatchingPredicate != null) {
+            return cachedMatchingPredicate;
+        }
+
+        final MatchingStrategy matchingStrategy = getMatchingStrategy(context);
+
+        switch (matchingStrategy) {
+            case ENUMERATION:
+                return determineMatchingPredicateBasedOnEnumeration(context, flowFile);
+            case REGEX:
+                return determineMatchingPredicateBasedOnRegex(context, flowFile);
+            default:
+                throw new IllegalArgumentException(
+                        "Cannot determine matching predicate for unsupported strategy " + matchingStrategy + " !"
```

Comment:
Exclamation marks should not be used in messages.
```suggestion
                        "Cannot determine matching predicate for unsupported strategy [" + matchingStrategy + "]"
```